sw/airborne/firmwares/rotorcraft/oneloop/oneloop_andi.c

```diff
@@ -0,0 +1,1524 @@
+  old_time = new_time;
+  printf("This function is running at a dt=%f \n",new_time-old_time);
+  printf("Half loop = %d \n",half_loop);
+  // At beginnig of the loop: (1) Register Attitude, (2) Initialize gains of RM and EC, (3) Calculate Normalization of Actuators Signals, (4) Propagate Actuator Model, (5) Update effectiveness matrix
+  float_eulers_of_quat_zxy(&eulers_zxy, stateGetNedToBodyQuat_f());
+  init_controller();
+  calc_normalization();
+  get_act_state_oneloop();
+  sum_g1g2_1l();
+  
+  // If drone is not on the ground use incremental law
+  use_increment = 0.0;
+  bool  volando       = false;
+  if(in_flight) {
+    use_increment = 1.0;
+    volando       = true;
+    printf("I am in flight\n");
+  }
+  if (ONELOOP_ANDI_DEBUG_MODE) {
+    volando = false;
+  }
+  
+  // Update desired Heading (psi_des_rad) based on previous loop or changed setting
+  psi_des_rad = psi_des_deg * M_PI / 180.0;
+
+  // Register the state of the drone in the variables used in RM and EC
+  // (1) Attitude related
+  att_1l[0] = eulers_zxy.phi                        * use_increment;
+  att_1l[1] = eulers_zxy.theta                      * use_increment;
+  att_1l[2] = eulers_zxy.psi                        * use_increment;
+  oneloop_andi_propagate_filters();   //needs to be after update of attitude vector
+  att_d[0]  = rates_filt_fo[0].last_out             * use_increment;
+  att_d[1]  = rates_filt_fo[1].last_out             * use_increment;
+  att_d[2]  = rates_filt_fo[2].last_out             * use_increment;
+  att_2d[0] = ang_acc[0]                            * use_increment;
+  att_2d[1] = ang_acc[1]                            * use_increment;
+  att_2d[2] = ang_acc[2]                            * use_increment;
+  // (2) Position related
+  pos_1l[0] = stateGetPositionNed_f()->x;   
+  pos_1l[1] = stateGetPositionNed_f()->y;   
+  pos_1l[2] = stateGetPositionNed_f()->z;   
+  pos_d[0]  = stateGetSpeedNed_f()->x;      
+  pos_d[1]  = stateGetSpeedNed_f()->y;      
+  pos_d[2]  = stateGetSpeedNed_f()->z;      
+  pos_2d[0] = lin_acc[0];
+  pos_2d[1] = lin_acc[1];
+  pos_2d[2] = lin_acc[2];
+  
+  // Update the effectiveness matrix used in WLS
+  int8_t i;
+  for (i = 0; i < ANDI_OUTPUTS; i++) {
+    bwls_1l[i] = g1g2_1l[i];
+  }
+
+  // Run the Reference Model (RM)
+  oneloop_andi_RM(half_loop, PSA_des, rm_order_h, rm_order_v);
+  // Generate pseudo control for stabilization vector (nu) based on error controller
+  if(half_loop){
+    nu[0] = 0.0;
+    nu[1] = 0.0;
+    nu[2] = a_thrust;
+    nu[3] = ec_3rd(att_ref[0], att_d_ref[0], att_2d_ref[0], att_3d_ref[0], att_1l[0], att_d[0], att_2d[0], k_phi_e, k_p_e, k_pdot_e);
+    nu[4] = ec_3rd(att_ref[1], att_d_ref[1], att_2d_ref[1], att_3d_ref[1], att_1l[1], att_d[1], att_2d[1], k_theta_e, k_q_e, k_qdot_e);
+    nu[5] = ec_2rd(att_d_ref[2], att_2d_ref[2], att_3d_ref[2], att_d[2], att_2d[2], k_r_e, k_r_d_e);
+  }else{
+    float k1_att_e[3] = {k_phi_e,  k_theta_e, k_psi_e};
```

Comment:
Why do this every time? You could make an array of pointers instead. Or perhaps directly define those gains in an array for simplicity.

Reply:
Done